### DISCUSSION TITLE
How can a loader add files to be processed?

### DISCUSSION THREAD

**Discussion by fregante** [webpack/webpack#19089] [Q&A]
I've been spending days on this and the documentation is just not enough.

I'm writing a loader that takes the input json file (added via `entry`) and processes all of its listed files according to the rules.
I tried:
- `this.addDependency()`: no output
- `this.emitFile()`: it does not follow the `rules` array, for example the one specifying the output directory
- `this.resolve()`: no output
- `new RawSource()`: I don't understand why I'd have to read each file directly, I'm sure webpack can do this autonomously

This leaves me with returning a generated piece of virtual JS string that has some `import` statements. Is this the best way?

The docs mention that both this OR `this.resolve()` should work, but that doesn't seem to be the case. Initially I wrote a plugin but that's even less clear since it cannot return said virtual JS string. 

https://webpack.js.org/contribute/writing-a-loader/#module-dependencies

**Comment by alexander-akait** [comment]
Sorry for delay, still actual? What do you want? Do you want execute file using webpack itself? Or do you want to bundle and inject? If you want just add in module graph just generate `import`/`require`/`new URL()` in sour code where you need it, no need to use entrypoint, because it is chunk, but you want to handle it as a module

**Reply by fregante** [reply]
I have a manifest.json with a list of files. The list of files needs to be treated as a list of entry points. One bundle per file listed. 

### input

manifest.json

```js
{entries: [
  "main.js",
  "other.js",
  "style.css"
]}
```

### output

- dist/main.js + its own chunks
- dist/other.js + its own chunks
- dist/style.css + its own chunks 

And the harder part would be to also output the manifest.json with the generated names, without hardcoding "dist/":

```js
{entries: [
  "dist/main.js",
  "dist/other.js",
  "dist/style.css"
]}
```

**Reply by alexander-akait** [reply]
Why do not read it in `webpack.config.js` and avoid extra complex things, anyway you can use create an own loader - read JSON file, use `this.addDependency(absolutePathToFile)` for the each file, so webpack will recompile on changes and generate code:
```
import * as ident_1 from "./main.js";
import * as ident_2 from "./other.js";
import * as ident_3 from "./style.css"
```

**Reply by fregante** [reply]
That's what I'm already doing, but:

- the manifest is not part of the build, so if I change it I have to restart webpack
- the manifest is not output with the generated files, so I have to carefully create a (static) manifest matching the output files rather than the source files
  - this means that if my webpack config is changed, the manifest will point to non-existing files

Also:

- I already tried this.addDependency() and it didn't work. Unfortunately there are no examples at all on the relevant documentation page, so I have no idea of how to use it properly 
- what's that code? Does that generate the 3 files I suggested in dist?

Mostly the reason to use this was because some of the entries are supposed to also be added dynamically, but I already failed to created a loader/plugin for the static manifest.json (which, by the way, is a web extension manifest)

**Reply by fregante** [reply]
What I'm curious about is why there isn't an API that I can call from a plugin that looks like this:

```JS
this.addEntry(path)
```

And works *exactly* as adding it to the `entry` array. 

And then after compilation:

```JS
const {distPath} = this.getEntry(path)
``` 

So that I can build the manifest *from the entry file list*

**Reply by alexander-akait** [reply]
Loader can't add entrypoint, loader should transform code and only, you can use this https://webpack.js.org/configuration/entry-context/#dynamic-entry if you need dynamic entry point

**Reply by alexander-akait** [reply]
Also you can just use plugin - https://github.com/webpack/webpack/blob/main/lib/DynamicEntryPlugin.js and using `entry` as a function where you parse/get and return entries
